### DISCUSSION TITLE
[Flutter] How to reduce time required to generate files with build_runner

### DISCUSSION THREAD

**Discussion by prolificcoder** [MobileNativeFoundation/discussions#200] [General]
Our flutter app is becoming huge about ~40 developers and 3000 tests and one aspect that is particularly annoying is generated files. 
Mockito -> *.mock.dart
Freezed -> *.freezed.dart
JSON Serializer -> *.g.dart
and some retrofit etc. 

Time to generate all these files
 -> linux 4g boxes: fails with `pub finished with exit code -9`
 -> linux 14g boxes: 16 mins  (CI box)
 -> m1 mac: 8 mins 
 -> mac-ec2: 13 mins (mac CI box)
 
 This seems very excessive and our app is only growing. Are other large flutter apps seeing similar times, are there any strategies to reduce this?

**Comment by chris-wickens** [comment]
Couple of suggestions:

1. Replace mockito with mocktail, which doesn't use code generation
1. Limit the code generation to run only in folders where your freezed files are defined. IIRC this can be done in analysis_options.yaml with wildcards, I can dig out more info if you need.

Ultimately code generation is a pain Flutter devs will probably have to live with until static metaprogramming lands, and we can only mitigate in the meantime.

**Reply by prolificcoder** [reply]
Thanks for the suggestions
1. We can also use mockito without the generated mocks. But changing all these tests now would be a huge task. 
2. Yeah looking at using `generate_for` in build.yaml. It gave some marginal improvements. 

I think a bigger problem is that we have generated files from freezed, json serializer, retrofit and mockito. Different developers started using them at different times and we didn't really realize the scale of the problem because we were also checking in these generated files and only occasionally re-generate them locally.

**Reply by chris-wickens** [reply]
Yep sorry, generate_for in build.yaml is what I meant. We managed to shave 30-40 seconds off the generation time, but YMMV depending on your directory structure. We check our generated files into source control also.

**Comment by prolificcoder** [comment]
A late update on this, can expand if there is interest but here are the steps we have taken broadly

1. Changed all the files that can generate mocks to add a suffix `.buildable` eg: bank_account.dart now becomes bank_account_buildable.dart -> This step was essential for flutter build gen to work under 5 minutes. This brought the total time for generated from 12 to ~5 mins. 
2. Add a build step in CI and locally to generate build files
3. Removed all generated files from source control 
4. Additionally, we created a setup step in CI that generates these build artifacts and passes them down the pipeline. 

Now generated files are bit more manageable.
